### ISSUE TITLE
Why init distance to 0?

### ISSUE THREAD

**Issue by MoYummy** [recastnavigation/recastnavigation#345]
https://github.com/recastnavigation/recastnavigation/blob/15ebb8bd25e952215d2ff140d5ec01fa50294275/Recast/Source/RecastMeshDetail.cpp#L566

**Comment by MoYummy** [comment]
~~~
	float dmin = 0;
	for (int i = 0; i < nhull; i++)
	{
		int pi = prev(i, nhull);
		int ni = next(i, nhull);
		const float* pv = &verts[hull[pi]*3];
		const float* cv = &verts[hull[i]*3];
		const float* nv = &verts[hull[ni]*3];
		const float d = vdist2(pv,cv) + vdist2(cv,nv) + vdist2(nv,pv);
		if (d < dmin)
		{
			start = i;
			left = ni;
			right = pi;
			dmin = d;
		}
	}
~~~

This will always find nothing, right?

**Comment by jakobbotsch** [comment]
Yes, that definitely looks very strange. It was changed here: https://github.com/recastnavigation/recastnavigation/commit/129efa5833f3ff8d4d54bf9d5b181555d486312b#diff-8242ad66e6c63acba9f3cac6b23b12f6R526

That function only triangulates convex polys and the result is correct anyway, but may be suboptimal compared to what was intended. It should be simple to change it to `FLT_MAX`, though we should check the impact on the triangulation of the detail mesh.

**Comment by MoYummy** [comment]
`FLT_MAX` is necessary but besides that I would suggest another improvement.

![image](https://user-images.githubusercontent.com/8344384/44512102-159e7900-a6ec-11e8-97fc-1b0581a2ca40.png)
![image](https://user-images.githubusercontent.com/8344384/44511680-e50a0f80-a6ea-11e8-9c34-5cdf9b684ccb.png)

As you can see from the image, the triangulation starts from the blue vertex (second smallest `y`).
However, at this point, all sample vertices are on edges, so the starting triangle is actually a line but with smallest perimeter. This is not good, right?
Therefore, I suggest looping over only the first `nin` of `verts` instead of all, which is vertices from original polygons, as the center vertex of the starting triangle.

**Comment by jakobbotsch** [comment]
I have not looked at this code for a while, but I think that will not be correct. Only the first `nin` of `verts` are original vertices. The triangulation happens over `hull` (indices into `verts`).
Instead we can add a check for collinearity, or mark internal samples as such (eg. with bitwise complement) and skip those.

**Comment by memononen** [comment]
The corrections from @MoYummy sound correct to me, as well as the skip from @jakobbotsch.

**Comment by jakobbotsch** [comment]
Great! Do you want to submit a PR with those changes @MoYummy?

**Comment by jakobbotsch** [comment]
Fixed by #346
